include/NovelRT.h

```diff
@@ -81,167 +84,285 @@
- */
-namespace NovelRT {
-  typedef class Atom Atom;
-  typedef class DebugService DebugService;
-  typedef class LoggingService LoggingService;
-  typedef class NovelRunner NovelRunner;
-}
-/**
- * @brief Contains animation features.
- */
-namespace NovelRT::Animation {
-  typedef class SpriteAnimator SpriteAnimator;
-}
-/***
- * @brief Contains audio features, such as playing audio, and managing audio resources.
- */
-namespace NovelRT::Audio {
-  typedef std::vector<ALuint> SoundBank;
-  typedef std::vector<ALuint> MusicBank;
-  typedef class AudioService AudioService;
-}
-/**
- * @brief Contains features to interop with the .NET Core runtime.
- */
-namespace NovelRT::DotNet {
-  typedef class RuntimeService RuntimeService;
-}
-/**
- * @brief Contains graphics features, such as rendering, textures, cameras, etc.
- */
-namespace NovelRT::Graphics {
-  typedef class Texture Texture;
-  typedef class BasicFillRect BasicFillRect;
-  typedef class Camera Camera;
-  typedef class ImageRect ImageRect;
-  typedef class RenderingService RenderingService;
-  typedef class RenderObject RenderObject;
-  typedef class TextRect TextRect;
-}
-/**
- * @brief Contains bindings for Ink.
- */
-namespace NovelRT::Ink {
-  typedef class InkService InkService;
-  typedef class Story Story;
-}
-/**
- * @brief Contains input features, such as keyboard and mouse interactivity.
- */
-namespace NovelRT::Input {
-  typedef class BasicInteractionRect BasicInteractionRect;
-  typedef class InteractionObject InteractionObject;
-  typedef class InteractionService InteractionService;
-}
-/**
- * @brief Contains scene graph features.
- */
-namespace NovelRT::SceneGraph {
-  typedef class QuadTreeNode QuadTreeNode;
-  typedef class QuadTreeScenePoint QuadTreeScenePoint;
-  typedef class RenderObjectNode RenderObjectNode;
-  typedef class Scene Scene;
-  typedef class SceneNode SceneNode;
-}
-/**
- * @brief Contains timing features, such as timers.
- */
-namespace NovelRT::Timing {
-  typedef class StepTimer StepTimer;
-}
-/**
- * @brief Contains windowing features.
- */
-namespace NovelRT::Windowing {
-  typedef class WindowingService WindowingService;
-}
+#if defined(__cplusplus) && !defined(NOVELRT_C_API)
+  /**
+   * @brief NovelRT is a cross-platform, flexible Visual Novel and 2D game engine.
+   * It is aimed at designers and developers alike, however many of the designer tools and features we have on our roadmap have yet to be implemented.
+   */
+  namespace NovelRT {
+    typedef class Atom Atom;
+    typedef class DebugService DebugService;
+    typedef class LoggingService LoggingService;
+    typedef class NovelRunner NovelRunner;
+  }
+#endif // __cplusplus
+
+#if defined(__cplusplus) && !defined(NOVELRT_C_API)
+  /**
+   * @brief Contains animation features.
+   */
+  namespace NovelRT::Animation {
+    typedef class SpriteAnimator SpriteAnimator;
+  }
+#else // !__cplusplus
+  #include "NovelRT.Interop/Animation/NrtAnimationTypedefs.h"
```

Comment:
IIRC, this was done due to some of the Interop headers requiring a definition for a type that was used in multiple areas, causing redefinitions to occur.

If this is how I'm imagining it, it _should_ be fine to make single file headers for Interop except the fact that we'll have multiple classes' interop versions falling into one file.

Would you be able to confirm? And if true, then @RubyNova do you have objections? (I don't, since it actually makes more sense in my head now than it did before lol)

Reply:
That sounds about right. Vulkan handles this by declaring the handles (typedefs) all up front at the top of `vulkan_core.h` (https://github.com/KhronosGroup/Vulkan-Headers/blob/master/include/vulkan/vulkan_core.h#L62) and then defining everything else below it.

That's basically what these `#include` in `NovelRT.h` are doing. It's ensuring all typedefs are defined before the main headers are included and then the headers defining the methods and such are below that.
It's basically required for something like ClangSharp because otherwise the header passed into ClangSharp has to itself ensure its doing the right includes and typedefs in the right order; vs just including `NovelRT.h` and traversing the files that define the types you want.

Reply:
Like the other convo - handle in separate PR? Or handle in this one and adjust the bindings appropriately?